### ISSUE TITLE
Diverse questions relative a cette contribution Kaggle

### ISSUE THREAD

**Issue by bourbe** [Laurae2/Santander#1]
Bonjour Laurae,

   Je m'adresse a toi en français car j'ai vu ton profil kaggle que tu étais en île de France. Moi aussi. 
   Je suis un nouveau passionné de datascience et j’étudie des codes comme le tien afin d'apprendre. 
   Merci de partager ainsi votre connaissance. J'ai donc qques questions concernant ton code si tu veux bien y répondre. 
   Je suis aussi dispo par tout autre moyen de communication que tu souhaiterai.

Question1 --------------------------------------------------------------------------------------
- Est ce que ton algorithme blend_proba (Stacker.py) prend en compte le cas de la multiclassification ?
- Au sein de cet algorithme on a ceci a adapter en fonction de la métrique

``` sh
 for i in range(len(fold_preds)): 
        if fold_preds[i]<0.0: #adapt to metric  
            fold_preds[i] = 0.0 #adapt to metric
        if fold_preds[i]>1.0: #adapt to metric 
            fold_preds[i] = 1.0 #adapt to metric 
```

 N'est ce pas plutôt a adapter en fonction en fonction du fait qu'on fait une classification binaire, multiclassification ou régression ?

Ce que je comprend la c'est que vu que tu va utiliser des "regressor" pour faire une classification binaire, tu corrige les sorties de la classification binaire de cette façon.

Pour moi sur un problème de régression ce ne serait pas nécessaire et que faire sur un problème de multiclassification ?
- Connais tu ce programme : https://github.com/MLWave/Kaggle-Ensemble-Guide/blob/master/blend_proba.py

Il fait la meme chose que le tien mais j'arrive pas a le faire fonctionner sous Python 3.5 : http://stackoverflow.com/questions/39687959/error-when-converting-form-python-2-to-python-3

Question2 --------------------------------------------------------------------------------------
- As tu des références sur cette technique qui te permet d'utiliser des regressor comme classifier stp ?

Question3 --------------------------------------------------------------------------------------
- J'ai regardé sur le net pour voir si il existait des stacking de plus de 3 levels et je n'en ai pas trouvé d'ou ma question pourquoi on ne va pas sur plus de 3 niveaux avec du stacking ? des problèmes de sur apprentissage ? Si oui pourquoi ?

Question4 --------------------------------------------------------------------------------------

---

``` sh
def log_loss_func(weights):
    ''' scipy minimize will pass the weights as a numpy array '''
    final_prediction = 0
    for weight, prediction in zip(weights, predictions):
            final_prediction += weight*prediction

    return log_loss(Y_testing, final_prediction) 


#the algorithms need a starting value, right not we chose 0.5 for all weights
#its better to choose many random starting points and run minimize a few times
starting_values = np.ones(len(predictions))/(len(predictions))

#adding constraints  and a different solver 
cons = ({'type':'eq','fun':lambda w: 1-sum(w)})

#our weights are bound between 0 and 1
bounds = [(0,1)]*len(predictions)

res = minimize(log_loss_func, starting_values, method='SLSQP', bounds=bounds, constraints=cons)
#res = minimize(log_loss_func, starting_values, method='SLSQP')


print('Ensamble Score: {best_score}'.format(best_score=res['fun']))
print('Best Weights: {weights}'.format(weights=res['x'])) 
```

---

Ton bout de code ci dessus permet de calculer les poids en minimisant le log_loss et résoud le problème d'optimisation ci dessous:

``` sh
op f(x)
    x =[a,b,c,......,n]
    0 <= xi <=1
    sum(x) = 1 
```

Mon problème concerne le cas ou    on veux utiliser le roc_auc_score et donc l'optimiser et non le minimiser. comment traduire le fait qu'on veuille 
optimiser et non minimiser

J'ai assayé les options suivantes et ce qui est surprenant j'ai la même sortie en terme de poids:
- Essai_1

``` sh
def roc_func(weights):
    ''' scipy minimize will pass the weights as a numpy array '''
    final_prediction = 0
    for weight, prediction in zip(weights, predictions):
            final_prediction += weight*prediction

    return roc_auc_score(Y_testing, final_prediction)
```
- Essai_2  (en inversant la fonction précédente et en appliquant minimize, je me dit qu'on fait une maximisation (comme avec hyperopt))

``` sh
def roc_func(weights):
    ''' scipy minimize will pass the weights as a numpy array '''
    final_prediction = 0
    for weight, prediction in zip(weights, predictions):
            final_prediction += weight*prediction

    return - roc_auc_score(Y_testing, final_prediction)
```
- Essai_3  

``` sh
def roc_func(weights):
    ''' scipy minimize will pass the weights as a numpy array '''
    final_prediction = 0
    for weight, prediction in zip(weights, predictions):
            final_prediction += weight*prediction

    return 1 - roc_auc_score(Y_testing, final_prediction)
```

**Comment by Laurae2** [comment]
> Est ce que ton algorithme blend_proba (Stacker.py) prend en compte le cas de la multiclassification ?

Non, ça n'a été prévu que pour le cas binaire.

> Au sein de cet algorithme on a ceci a adapter en fonction de la métrique. N'est ce pas plutôt a adapter en fonction en fonction du fait qu'on fait une classification binaire, multiclassification ou régression ?

Le script n'est pas adapté pour la classification multiclasse. Il faut le ré-adapter.

> Pour moi sur un problème de régression ce ne serait pas nécessaire et que faire sur un problème de multiclassification ?

Si tu souhaites utiliser les outputs immédiatement sans borner, tu ne pourras pas (ex : si tu calcules le logloss entre la prédiction -1.5 et la vérité 0, pas possible. De plus, ça permet d'éviter que des variables soient affectées trop par des erreurs dans les features (genre une valeur -2147483647 hasardeuse dans une feature quand y a que des 0 et des 1 par exemple).

> https://github.com/MLWave/Kaggle-Ensemble-Guide/blob/master/blend_proba.py

Oui, faut réadapter pour Python 3.5 avec les changements adéquats. Si tu n'arrives pas à réadapter le script, utilises Python 2.7.

> As tu des références sur cette technique qui te permet d'utiliser des regressor comme classifier stp ?

Quasiment tous les modèles de régression sont des classifieurs. Par exemple, une régresion logisitique est un classifieur dont la sortie est une sigmoïde qui lui permet d'être utiliser comme classifieur.

> J'ai regardé sur le net pour voir si il existait des stacking de plus de 3 levels et je n'en ai pas trouvé d'ou ma question pourquoi on ne va pas sur plus de 3 niveaux avec du stacking ? des problèmes de sur apprentissage ? Si oui pourquoi ?

Il faut le coder soi-même. De plus ici, ce n'est pas un modèle de stacking complet en soit, c'est un prédicteur out-of-fold pour que tous les éléments du stacker soient prêts, mis à part le modèle d'apprentissage (il ne fait pas le stacking seul, il manque le(s) modèle(s) pour utiliser les out-of-fold et apprendre). Entre autres, un modèle à 3 niveaux complets est quelque chose qui doit être mis en place soigneusement et son architecture dépendra intégralement de tes nécessités. Un script pour du stacking à 3 niveaux sur une dataset ne pourra pas être utilisé de manière efficace sur un autre dataset. Même pour 2 niveaux, choisir le setup est compliqué.

Tu peux aller à 3+ niveaux, mais le gain est quasiment nul.

> Mon problème concerne le cas ou on veux utiliser le roc_auc_score et donc l'optimiser et non le minimiser. comment traduire le fait qu'on veuille optimiser et non minimiser

1 - AUC car max(AUC) = 1 et min(AUC) = 0 si tu cherches à minimiser l'objectif.

**Comment by bourbe** [comment]
Merci pour ton retour rapide. Cela éclairci bien des points. Pour le dernier point si je comprends bien vu que je cherche a minimiser l'objectif je dois utiliser 1 - auc qui aura pour resultat global de maximiser l'auc.

derniere demande, je susi en train de construire une architecture  sur trois niveaux qui m'a demandé beaucoup, beacoup de lecture afin d'y arriver. Ca te dirais de me donner ton avis stp ?

Merci bien encore

**Comment by Laurae2** [comment]
> Pour le dernier point si je comprends bien vu que je cherche a minimiser l'objectif je dois utiliser 1 - auc qui aura pour resultat global de maximiser l'auc.

Vu que tu va minimiser 1 - AUC, quand tu arriveras à 0 ça voudra dire que t'auras atteint AUC = 1.

> derniere demande, je susi en train de construire une architecture sur trois niveaux qui m'a demandé beaucoup, beacoup de lecture afin d'y arriver. Ca te dirais de me donner ton avis stp ?

Postes le.

**Comment by bourbe** [comment]
https://anaconda.org/bourbe/bank_new_blend_proba_1/notebook

**Comment by bourbe** [comment]
https://github.com/bourbe/PML/blob/master/bank_new_blend_proba_1.ipynb

**Comment by bourbe** [comment]
Pour decrire au global de ce que j'ai compris du Stacking, Je cree des oop pour chacun des trois niveaux pour le train set et de predictions pour le test set

En passant d'un niveau à l'autre, on rajoute les oop comme feature au train et les prediction comme feature au test set.

Entre chaque niveau, je fais une evaluation pour voir ce que je gagne en calculant les poids 

pour la fin je fais un tuning sur un xgboost, un ridge classifier et un extratree et ce sera le dernier niveau que j'agregerai par des poids.

Qu'en penses tu ? ai enfin compris le Stacking ? 

j'utilise bien entendu ton code stacker.py pour calculer les oop et les predictions

Bien à toi

**Comment by bourbe** [comment]
oop = out of fold predictions

**Comment by bourbe** [comment]
Bonjour,

  Juste une petite relance cordiale  -:)

  https://github.com/bourbe/PML/blob/master/bank_new_blend_proba_1.ipynb

Bonne Journée

**Comment by Laurae2** [comment]
J'ai besoin de trouver du temps pour lire le code (y a ~10,000 lignes), c'est pas comme si ça me prendrait 5 minutes pour répondre.

En lisant rapidement, tout l'architecture me semble correcte (je fais confiance au nom des variables), à part pour les optimiseurs où tu n'as pas besoin d'utiliser de bornes sauf si tu comptes utiliser les prédictions pour un vrai usage réel (e.g tester tes prédictions, pas pour stacker/ensemble). Me semble pas voir de leakage.

En théorie (et ça a déjà été démontré) le fait de retirer les bornes [-Inf, Inf] aux optimiseurs pour un stack linéaire de modèle ça permet d'améliorer mieux la performance que s'il y a un bornage positif [0, +Inf]. En pratique, ça donne souvent des résultats meilleurs que le bornage positif. En tout cas, étonnant que tu aies quasiment les mêmes coefficients trouvés par l'optimiseur (j'utilise aucun solveur en Python, donc j'ai pas l'expérience dessus). N'oublies pas d'ajuster l'initialisation à une initialisation aléatoire si c'est disponible (utiliser une initialisation identique à un optimiseur revient à lui dire que si la fonction de perte varie dans un sens, alors toutes les variables varient dans le même sens à la même intensité).

**Comment by bourbe** [comment]
Merci Laurae2 pour ton retour. 

Désolé pour mon empressement, je suis un peu comme un enfant devant tous ca. De prime a bord je suis ravi que tu dises que l'architecture est correcte ce qui signifie que je pense avoir enfin compris le stacking (j'ai longtemps cherché ...;) 
- Est ceci les bornes ? Si oui pourquoi ne pas avoir besoin car ils permettent de dire a l'optimiseur de nous retourner des poids entre 0 et 1 ?

``` sh
#our weights are bound between 0 and 1
bounds = [(0,1)]*len(predictions)
```
- Qu'est ce que le leakeage ? j'ai vu ca sur le net et je vais creuser : https://www.quora.com/Whats-data-leakage-in-data-science
- Je vais faire un test  concernant le bornage (retirer)
- Effectivement pour le fait que je trouve des coefficient identiques , je me pose toujours la question de savoir ce qui se passe. Avec le log_loss le comportement est different, j'ai des coefficients qui varient.
-  Je n'ai rien compris a ceci:  N'oublies pas d'ajuster l'initialisation à une initialisation aléatoire si c'est disponible (utiliser une initialisation identique à un optimiseur revient à lui dire que si la fonction de perte varie dans un sens, alors toutes les variables varient dans le même sens à la même intensité).

Je regarderai tous ces point en details ce week end et je te remercie encore pour ton aide et encore desolé pour le coup de pression. 

c'est vraiment interessant d'echanger avec un/une kaggle master

Merci encore :-)

**Comment by Laurae2** [comment]
> Est ceci les bornes ? Si oui pourquoi ne pas avoir besoin car ils permettent de dire a l'optimiseur de nous retourner des poids entre 0 et 1 ?

prédictions entre [0, 1] != poids entre [0, 1]

ça dépend comme j'ai dis de l'utilisation que tu fais de ta prédiction après.

> Qu'est ce que le leakeage ?

Quelque chose qui est inhérent à une variable qui se reflète directement (et non pas indirectement) sur le label. Genre par exemple quand tu fais une étude épidémiologique comprennant des individus tracés sur plusieurs années, si tu laisses la personne en tant que feature dans le training, il est possible que tes modèles apprennent les individus au lieu d'apprendre à généraliser.

> Effectivement pour le fait que je trouve des coefficient identiques , je me pose toujours la question de savoir ce qui se passe. Avec le log_loss le comportement est different, j'ai des coefficients qui varient.

Ce sont deux choses non comparables. Un AUC élevé ne rime pas avec un logloss faible, et vice-versa. Tu peux très bien avoir AUC = 1 et logloss = 0.69 dans une dataset avec 50% de 0 et de 1.

> Je n'ai rien compris a ceci

En optimisation c'est la base : si tu initialises mal des paramètres, tu auras des résultats qui ne sont qu'un minimum local élevé si tu as une fonction non convexe. cf par exemple gradient descent à plusieurs variables. De plus, si tu donnes un poids identique à chaque variable d'emblée, il n'y aura aucun moyen de faire savoir à l'optimiseur si tel ou tel poids implique un changement de perte différent, puisqu'il n'y a aucun moyen de les différencier (à moins que l'optimiseur prenne en charge ce cas là).

P.S : créés un topic sur Kaggle ou sur stackoverflow à la place d'ici, car ça diverge sur des topics généraux qui n'ont plus grand chose à voir avec les scripts de la repo (questions sur des fondamentaux de Maths/ML).

**Comment by bourbe** [comment]
Bonjour,

  merci pour tous ceci.  Je crée un topic sur Kaggle. Tu peux clore cette issue

Bonne Journée
